### PR TITLE
Make Installing nearly instantaneous

### PR DESCRIPTION
**Pull request by Validark**
- Eliminate need for python script via universal Github Installer
- Make Install script change HttpEnabled back to what it was

Test code:
```lua
local h = game:GetService("HttpService") local e = h.HttpEnabled h.HttpEnabled = true loadstring(h:GetAsync("https://raw.githubusercontent.com/Validark/NevermoreEngine/patch-3/Install.lua"))(e)
```

### CODE REVIEW

File: Install.lua [hunk 1]
```diff
@@ -1,175 +1,275 @@
...
+	-- Installs Link into Parent
+	Link = Link:gsub("/$", "")
+
+	-- Extract Link Data
+	local Organization, Repository, Tree, ScriptName, ScriptClass
+	local Website, Directory = Link:match("^(https://[raw%.]*github[usercontent]*%.com/)(.+)")
+	Organization, Directory = (Directory or Link):match("^/?([%w-_%.]+)/?(.*)")
+	Repository, Directory = Directory:match("^([%w-_%.]+)/?(.*)")
+
+	if Website == "https://raw.githubusercontent.com/" then
```
Comment (Quenty) [inline Install.lua]: Code clarify:

Should break this if-statement up into several methods.

File: Install.lua [hunk 2]
```diff
@@ -1,175 +1,275 @@
...
+		if Directory then
```
Comment (Quenty) [inline Install.lua]: Can be a method called `Github:ExtractScriptData(Directory)` which returns a table with ScriptName/ScriptClass. Make it much more clear what you're doing.

File: Install.lua [hunk 3]
```diff
@@ -1,175 +1,275 @@
...
+			Tree, Directory = Directory:match("^([^/]+)/(.+)")
+			if Directory then
+				ScriptName, ScriptClass = Directory:match("([%w-_%%]+)%.?(%l*)%.lua$")
+			end
+		end		
+	elseif Directory then
+		local a, b = Directory:find("^[tb][rl][eo][eb]/[^/]+")
```
Comment (Quenty) [inline Install.lua]: Makes it hard to tell blob/tree is being extracted beforehand. Maybe extract `(%w+)/([^/]+)` and check the first word if it's tree/blob, and the second for the match using `ExtractScriptData` method suggested instead of repeated code.  

In reality this is to grab scriptdata from a directory, so probably break this up too.

File: Install.lua [hunk 4]
```diff
@@ -1,175 +1,275 @@
...
+			if Directory == "" then Directory = nil end
+			if Directory and Link:find("blob") then
+				ScriptName, ScriptClass = Directory:match("([%w-_%%]+)%.?(%l*)%.lua$")
+			end
+		else
+			Directory = nil
+		end
+	end
+
+	if not Website then Website = "https://github.com/" end
```
Comment (Quenty) [inline Install.lua]: Can replace with `Website = Website or "https://github.com"` pattern, otherwise probably tab correctly.

File: Install.lua [hunk 5]
```diff
@@ -1,175 +1,275 @@
...
+	Directory = Directory and ("/" .. Directory):gsub("^//", "/") or ""
+
+	-- Threads
+	local Routines = RoutineList or {false}
```
Comment (Quenty) [inline Install.lua]: Method shouldn't do both extraction and task scheduling like this, interloping with global variables

Comment (Quenty) [reply Install.lua]: Instead: split up into multiple variables/methods

File: Install.lua [hunk 6]
```diff
@@ -1,175 +1,275 @@
...
+	local Value = #Routines + 1
+	Routines[Value] = false
+
+	local Garbage
+
+	if ScriptName then
+		local Script = Instance.new(ScriptTypes[ScriptClass or "mod"])
```
Comment (Quenty) [inline Install.lua]: This should be in a method like `CreateScript()`

File: Install.lua [hunk 7]
```diff
@@ -1,175 +1,275 @@
...
-MakeScript(ReplicatedStorage, "App/NevermoreEngine.lua")
+		if MainExists then Directory = Directory + 1 end -- :gsub("[^/]+$", "", 1) end			
+		local Count = 0
+
+		local function LocateFolder(FolderName)
+			Count = Count + 1
+			if Count > Directory then
+				Directory = Directory + 1
+				if (Parent and Parent.Name) ~= FolderName and "Modules" ~= FolderName then
+					if FolderName == "App" and Link:find("/Quenty/NevermoreEngine/", 1, true) then FolderName = "ReplicatedStorage" end
```
Comment (Quenty) [inline Install.lua]: Note for the source of this file, should avoid hardcoding this stuff in. OK in an installer. Bit weird in a generic installer module.

File: Install.lua [hunk 8]
```diff
@@ -1,175 +1,275 @@
...
+		DataSources[Object] = Link
+		return Object
+	end
 end
 
-HttpService.HttpEnabled = WasHttpEnabled
-ProgressText.Text = "Done!"
-done = true
+GitHub:Install("/Quenty/NevermoreEngine", game:GetService("ServerScriptService")).Name = "Nevermore"
+HttpService.HttpEnabled = ...
```
Comment (Quenty) [inline Install.lua]: Should leave a comment why we're doing this/usecase, incase someone tries to install this the other way.

File: Install.lua [hunk 9]
```diff
@@ -1,175 +1,275 @@
...
-local Paths = {}
-for ScriptPath in (LoadURL("Modules/ModuleList.txt")):gmatch("[^\r\n]+") do
-	Paths[#Paths+1] = ScriptPath
-	print""
+		for a = 2, ScriptCount do
+			local Link = Scripts[a]
+			local ScriptName, ScriptClass = Link:match("([%w-_%%]+)%.?(%l*)%.lua$")
+			local Script = Instance.new(ScriptTypes[ScriptClass or TypesSpecified and "" or "mod"], Parent)
+			Script.Name = ScriptName:gsub("Library$", "", 1):gsub("%%(%x%x)", UrlDecode)
+			coroutine.resume(coroutine.create(GiveSourceToScript), "https://raw.githubusercontent.com" .. Link:gsub("(/[^/]+/[^/]+/)blob/", "%1", 1), Script)
```
Comment (Quenty) [inline Install.lua]: coroutine passing like this is hard to follow, need to be careful in certain cases. DownloadScriptSource maybe better name? Concatenation of link here hard to follow.

**Comment by Quenty** [comment]
Oh yeah, also please remove the python installer and the install list! We won't need it anymore.

**Comment by Validark** [comment]
- Detects previous install
- Updates existing versions.

**Comment by Quenty** [comment]
Not perfectly clean, but we'll take this. Nice job!

**Comment by Quenty** [comment]
Before we can merge, you'll need to merge in the latest version than `git rm ModuleList.txt` and the `CompileBuilder.py`. Sorry!

**Comment by Validark** [comment]
You'll have to delete the ModuleList.txt yourself.

**Comment by Validark** [comment]
Ready for merging.

**Comment by Quenty** [comment]
Merged! Thank you!
